packages/block-library/src/button/edit.native.js

```diff
@@ -0,0 +1,201 @@
+/**
+ * External dependencies
+ */
+import { View, Dimensions, InteractionManager } from 'react-native';
+/**
+ * WordPress dependencies
+ */
+import {
+	withInstanceId,
+	compose,
+} from '@wordpress/compose';
+import { __ } from '@wordpress/i18n';
+import {
+	RichText,
+	withColors,
+	InspectorControls,
+	__experimentalUseGradient,
+} from '@wordpress/block-editor';
+import {
+	TextControl,
+	ToggleControl,
+	PanelBody,
+	RangeControl,
+	MissingControl,
+	NotificationSheet,
+} from '@wordpress/components';
+import {
+	useCallback,
+	useState,
+} from '@wordpress/element';
+import { withDispatch } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import richTextStyle from './richText.scss';
+import styles from './editor.scss';
+import RichTextWrapper from './richTextWrapper';
+
+const NEW_TAB_REL = 'noreferrer noopener';
+const INITIAL_BORDER_RADIUS = 4;
+const MIN_BORDER_RADIUS_VALUE = 0;
+const MAX_BORDER_RADIUS_VALUE = 50;
+const BORDER_WIDTH = 1;
+const BLOCK_SPACING = 4;
+
+function ButtonEdit( { attributes, setAttributes, backgroundColor, textColor, isSelected, closeGeneralSidebar } ) {
+	const {
+		placeholder,
+		text,
+		borderRadius,
+		url,
+		linkTarget,
+		rel,
+	} = attributes;
+
+	const [ isFocused, setRichTextFocus ] = useState( true );
+	const [ showHelp, setShowHelp ] = useState( false );
+
+	const borderRadiusValue = borderRadius !== undefined ? borderRadius : INITIAL_BORDER_RADIUS;
+	const mainColor = backgroundColor.color || '#595959';
+	const outlineBorderRadius = borderRadiusValue > 0 ? borderRadiusValue + BLOCK_SPACING + BORDER_WIDTH : 0;
+
+	const onToggleOpenInNewTab = useCallback(
+		( value ) => {
+			const newLinkTarget = value ? '_blank' : undefined;
+
+			let updatedRel = rel;
+			if ( newLinkTarget && ! rel ) {
+				updatedRel = NEW_TAB_REL;
+			} else if ( ! newLinkTarget && rel === NEW_TAB_REL ) {
+				updatedRel = undefined;
+			}
+
+			setAttributes( {
+				linkTarget: newLinkTarget,
+				rel: updatedRel,
+			} );
+		},
+		[ rel, setAttributes ]
+	);
+
+	const setBorderRadius = ( value ) => {
+		setAttributes( {
+			borderRadius: value,
+		} );
+	};
+
+	const toggleShowNoticationSheet = () => {
+		setShowHelp( ! showHelp );
+	};
+
+	const openNotificationSheet = () => {
+		closeGeneralSidebar();
+		InteractionManager.runAfterInteractions( () => {
+			toggleShowNoticationSheet();
+		} );
+	};
+
+	const {
+		gradientValue,
+	} = __experimentalUseGradient();
+
+	// 2 * container padding (32) + 2 * rich text padding (32) + 2 * BLOCK_SPACING (8) + 2 * BORDER_WIDTH = 74
+	// 580 is a max width when screen has horizontal orientation. Value comes from `ReadableContentView` styles.
+	const maxWidth = Math.min( Dimensions.get( 'window' ).width - 74, 580 - 74 );
```

Comment:
this looks super fragile. can't we use onLayout to detect container view's width instead?

Reply:
Could you please elaborate more on how do you see it?

Reply:
Generally, I think that `onLayout` will not help there because we need to know the `maxWidth` before the `View` is start increasing.

Reply:
This is what I meant by fragile:

<img width="393" alt="Screen Shot 2019-12-12 at 20 25 06" src="https://user-images.githubusercontent.com/5032900/70734636-a9e78180-1d1d-11ea-8b85-f613df114ef6.png">

Demonstrated HTML:

```
<!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link">Button</a></div>
<!-- /wp:button -->

<!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:button -->
<div class="wp-block-button"><a class="wp-block-button__link">Button 2</a></div>
<!-- /wp:button --></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group -->

<!-- wp:paragraph -->
<p></p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p></p>
<!-- /wp:paragraph -->
```

Current maxWidth calculation takes the window width as the base. It won't work for several cases and it'll need constant update as we change block container margins/paddings. What I am trying to say is that we can add one more wrapper View like:

```
<View onLayout={ onLayout } style={ { flex:1 } }>
```

That way we'll have chance to get our container view width:

```
	const onLayout = ( event ) => {
		const { width, height } = event.nativeEvent.layout;
		console.log(`width: ${ width } height: ${height}`);
	};
```

We can start with a default maxWidth and once onLayout is fired we can update the maxWidth of our RichText. That was my idea. Does that sound good?

Reply:
@pinar Sounds really good, I totally forgot about case where `Button` is nested. Will refactor it according to your suggestion. Thanks!

Reply:
Fixed!